Source/DafnyLanguageServer/Workspace/DafnyDocument.cs

```diff
@@ -72,8 +79,23 @@ public record DafnyDocument(
 
     public int LinesCount => VerificationTree.Range.End.Line;
     public IVerificationProgressReporter? GutterProgressReporter { get; set; }
-    public ConcurrentStack<Counterexample>? CounterexamplesCollector { get; set; }
-    public ConcurrentDictionary<ImplementationId, ImplementationView>? ImplementationIdToViewCollector { get; set; }
+    public ConcurrentStack<Counterexample> Counterexamples { get; set; } = new();
+    public ConcurrentDictionary<ImplementationId, ImplementationView> ImplementationIdToView { get; set; } = new();
+
+    /// <summary>
+    /// Creates a clone of the DafnyDocument
+    /// </summary>
+    public DafnyDocument Snapshot() {
+      var result = new DafnyDocument(TextDocumentItem, ParseAndResolutionDiagnostics, CanDoVerification, GhostDiagnostics,
+        Program, SymbolTable, WasResolved, LoadCanceled)
+      {
+        VerificationTree = VerificationTree,
+        Counterexamples = new(Counterexamples),
```

Comment:
Is this thread-safe?  I would have thought you'd need a lock around this to make sure that the various data structures are in sync with each other.

Reply:
This is only called inside a lock now.